### ISSUE TITLE
Android CMake toolchain does not define CMAKE_ASM_COMPILER

### ISSUE THREAD

**Issue by DoDoENT** [android-ndk/ndk#234]
I am trying to build OpenCV using new toolchain that comes with r13b, instead of using theirs. When building ARMv7 with NEON version of libpng, there is compile error when compiling .S file containing neon assembly code, because ninja/make invokes clang as assembler and does not give it `-mfpu=neon` flag.

Apparently, adding following line to android.toolchain.cmake fixes the problem:

    set( CMAKE_ASM_COMPILER "${ANDROID_TOOLCHAIN_PREFIX}gcc${ANDROID_TOOLCHAIN_SUFFIX}" )

**Comment by DanAlbert** [comment]
We might want to fix this by setting `CMAKE_ASM_COMPILER` to `clang -mfpu=neon` instead. I'll take a look at what we do in ndk-build and match that.

**Comment by DoDoENT** [comment]
@DanAlbert, be careful with `clang -mfpu=neon`! Think that some `.S` files might contain x86, x86_64 or arm64 assembly. Of course, it is up to CMake script to load correct `.S` file based on android ABI, but still compiling x86 SSE assembly with `-mfpu=neon` won't work.

**Comment by DanAlbert** [comment]
> @DanAlbert, be careful with clang -mfpu=neon! Think that some .S files might contain x86, x86_64 or arm64 assembly.

Conditionally on the target ABI being ARM with neon, of course :)

**Comment by DanAlbert** [comment]
I couldn't repro this. Here's my test case: https://android-review.googlesource.com/c/304181/

It does fail if I don't pass `-DANDROID_ARM_NEON=ON` when running cmake. Is libpng not doing that?

**Comment by DoDoENT** [comment]
I was building with `-DANDROID_ABI=armeabi-v7a` `-DANDROID_ARM_NEON=ON` and `-DENABLE_NEON=ON` and was using Ninja Generator (the entire cmake invocation line is below).

The latter is the variable from [libpng's fork in OpenCV](https://github.com/opencv/opencv/blob/3.1.0/3rdparty/libpng/CMakeLists.txt#L6) (note that I am building from tag v3.1.0, not the latest master).

Here is the entire cmake invocation shell script:

```
export ANDROID_NDK=~/android-sdks/ndk-bundle
cmake -GNinja -DANDROID_TOOLCHAIN_NAME=clang -DANDROID_ABI=armeabi-v7a -DANDROID_ARM_NEON=ON -DENABLE_NEON=ON -DANDROID_STL=c++_static -DANDROID_CPP_FEATURES="rtti exceptions" -DANDROID_PLATFORM=android-9 -DBUILD_ANDROID_EXAMPLES=OFF -DBUILD_DOCS=OFF -DBUILD_FAT_JAVA_LIB=OFF -DBUILD_JASPER=OFF -DBUILD_OPENEXR=OFF -DBUILD_PACKAGE=OFF -DBUILD_PERF_TESTS=OFF -DBUILD_TESTS=OFF -DBUILD_TIFF=ON -DBUILD_WITH_DEBUG_INFO=OFF -DBUILD_opencv_apps=OFF -DBUILD_opencv_java=OFF -DBUILD_opencv_python2=OFF -DBUILD_opencv_world=OFF -DCMAKE_C_FLAGS_RELEASE="-Os -DNDEBUG -fvisibility=hidden -ffunction-sections -fstack-protector-all" -DCMAKE_CXX_FLAGS_RELEASE="-Os -DNDEBUG -fvisibility=hidden -ffunction-sections -fstack-protector-all -fvisibility-inlines-hidden" -DENABLE_PRECOMPILED_HEADERS=OFF -DWITH_EIGEN=OFF -DWITH_JASPER=OFF -DWITH_OPENEXR=OFF -DWITH_TIFF=ON -DWITH_TBB=ON -DWITH_CUDA=OFF -DWITH_CUFFT=OFF -DWITH_WEBP=OFF -DCMAKE_BUILD_WITH_INSTALL_RPATH=ON -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK/build/cmake/android.toolchain.cmake $1
```

**Comment by DanAlbert** [comment]
Just tried exactly that and everything worked. Then I tried with r13 and yeah, broken.

Looked at the git long and remembered that someone fixed our asm stuff in cmake about a month ago: https://android-review.googlesource.com/c/291588/

Looks like this is already fixed for r14 then :)

**Comment by DanAlbert** [comment]
@DoDoENT: FYI, I just tried building that project with `-DANDROID_UNIFIED_HEADERS=ON`, and it looks like the way I've implemented that for cmake doesn't work with `check_include_file`. https://github.com/android-ndk/ndk/issues/238

There were a handful of other issues too (missing `#include <fcntl.h>` in most of the tif sources). Just a heads up that this probably won't work right away with r14 beta 1.

**Comment by DanAlbert** [comment]
Actually, now that I have a fix for the former, all that needs to be fixed is for the missing `#include`s to be added and FP16 support disabled (opencv tries to define functions with the same names as things in arm_neon.h, but in Clang those are macros rather than functions).

**Comment by leugenea** [comment]
If anyone's interested…
As mentioned above, bug with Clang, ASM and NEON fixed in master, so taking android.toolchain.cmake (you can grab one here: https://android.googlesource.com/platform/ndk.git/+/master/build/cmake/) from master fixed OpenCV 3.1.0 build. Also, don't forget to add -DENABLE_NEON=ON option to CMake (for libpng).
